File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.js
function solution(numbers) {
    let answer = new Array(numbers.length).fill(-1);
    let maxArr = [];

    for (let i = numbers.length - 2; i >= 0; i--) {
        if (numbers[i] < numbers[i + 1]) {
            answer[i] = numbers[i + 1];
            maxArr.push(numbers[i + 1]);
        } else {
            while (maxArr.length) {
                if (numbers[i] < maxArr[maxArr.length - 1]) {
                    answer[i] = maxArr[maxArr.length - 1];
                    break;
                } else {
                    maxArr.pop();
                }
            }
        }
    }

    return answer;
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];
        Stack<Integer> stack = new Stack<>();
        
        answer[answer.length-1] = -1;

        for(int i=numbers.length-2; i>=0; i--) {
            if(numbers[i] < numbers[i+1]) {
                answer[i] = numbers[i+1];
                stack.push(numbers[i+1]);
            } else {
                while(!stack.isEmpty()) {
                    if(numbers[i] < stack.peek()) {
                        answer[i] = stack.peek();
                        break;
                    } else {
                        stack.pop();
                    }
                }

                if(answer[i] == 0) {
                    answer[i] = -1;
                }
            }
        }

        return answer;
    }
}